src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs

```diff
@@ -0,0 +1,142 @@
+            if (context.TargetToken.RawKind == syntaxFacts.SyntaxKinds.DotToken)
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+
+            var typeInferenceService = context.GetLanguageService<ITypeInferenceService>();
+            var inferedTypes = typeInferenceService.InferTypes(context.SemanticModel, position, cancellationToken);
+
+            var hideAdvancedMembers = options.GetOption(RecommendationOptions.HideAdvancedMembers, context.SemanticModel.Language);
+
+            // We'll want to build a list of the actual enum members and all accessible instances of that enum, too
+            var result = inferedTypes
+                .Where(s => s.TypeKind == TypeKind.Enum)
+                .SelectMany(e => e.GetMembers().OfType<IFieldSymbol>())
+                .Where(f =>
+                    f.IsConst &&
+                    f.IsEditorBrowsable(hideAdvancedMembers, context.SemanticModel.Compilation))
+                .ToImmutableArray<ISymbol>();
+
+            return Task.FromResult(result);
+        }
+
+        protected override Task<ImmutableArray<ISymbol>> GetSymbolsAsync(
+            SyntaxContext context, int position, OptionSet options, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = context.GetLanguageService<ISyntaxFactsService>();
+            if (syntaxFacts.IsInNonUserCode(context.SyntaxTree, context.Position, cancellationToken))
+            {
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+            }
+
+            // This providers provides fully qualified names, eg "DayOfWeek.Monday"
+            // Don't run after dot because SymbolCompletionProvider will provide
+            // members in situations like Dim x = DayOfWeek.$$
+            if (context.TargetToken.RawKind == syntaxFacts.SyntaxKinds.DotToken)
+            {
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+            }
+
+            var typeInferenceService = context.GetLanguageService<ITypeInferenceService>();
+            var enumType = typeInferenceService.InferType(context.SemanticModel, position, objectAsDefault: true, cancellationToken: cancellationToken);
+
+            if (enumType?.TypeKind != TypeKind.Enum)
+            {
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+            }
+
+            var hideAdvancedMembers = options.GetOption(RecommendationOptions.HideAdvancedMembers, context.SemanticModel.Language);
+
+            var otherSymbols = context.SemanticModel.LookupSymbols(position).WhereAsArray(s =>
+                s.MatchesKind(SymbolKind.Field, SymbolKind.Local, SymbolKind.Parameter, SymbolKind.Property) &&
+                s.IsEditorBrowsable(hideAdvancedMembers, context.SemanticModel.Compilation));
+
+            var otherInstances = otherSymbols.WhereAsArray(s => Equals(enumType, s.GetSymbolType()));
+
+            return Task.FromResult(otherInstances.Concat(enumType));
```

Comment:
The behavior of this method is odd. It only returns the Enum, but not the enum values. For C# I would expect the exact opposite behavior (There are EnumAndCompletionListTagCompletionProvider.cs (and maybe AbstractRecommendationServiceBasedCompletionProvider.cs) that add the Enum type to the completion list).

I can't tell for VB if there is another provider that adds the Enum type.